src/Microsoft.VisualStudio.Threading.Shared/AsyncBarrier.cs

```diff
@@ -52,21 +50,33 @@ public AsyncBarrier(int participants)
         /// <returns>An awaitable.</returns>
         public Task SignalAndWait()
         {
-            var tcs = new TaskCompletionSource<bool>();
-            this.waiters.Push(tcs);
-            if (Interlocked.Decrement(ref this.remainingParticipants) == 0)
+            lock (this.waiters)
```

Comment:
That looks almost like what a lock does internally, except that you yield instead of block, which makes it functionality equivalent to `AsyncSemaphore`. :)
But my prior statement I think still stands: I don't want to make this more complex without evidence that contention is hurting us. Since we have no users at all...